File: src/go/tools.go
package geminisdk

import (
	"context"
	"encoding/json"
	"sync"
)

// ToolRegistry manages tool definitions and handlers
type ToolRegistry struct {
	tools    map[string]Tool
	handlers map[string]ToolHandler
	mu       sync.RWMutex
}

// NewToolRegistry creates a new tool registry
func NewToolRegistry() *ToolRegistry {
	return &ToolRegistry{
		tools:    make(map[string]Tool),
		handlers: make(map[string]ToolHandler),
	}
}

// Register adds a tool with its handler
func (r *ToolRegistry) Register(tool Tool, handler ToolHandler) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.tools[tool.Name] = tool
	r.handlers[tool.Name] = handler
}

// Tools returns all registered tools
func (r *ToolRegistry) Tools() []Tool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	
	result := make([]Tool, 0, len(r.tools))
	for _, tool := range r.tools {
		result = append(result, tool)
	}
	return result
}

// GetTool returns a tool by name
func (r *ToolRegistry) GetTool(name string) (Tool, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	tool, ok := r.tools[name]
	return tool, ok
}

// GetHandler returns a handler by name
func (r *ToolRegistry) GetHandler(name string) (ToolHandler, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	handler, ok := r.handlers[name]
	return handler, ok
}

// Execute runs a tool invocation
func (r *ToolRegistry) Execute(ctx context.Context, invocation ToolInvocation) ToolResult {
	r.mu.RLock()
	handler, ok := r.handlers[invocation.Name]
	r.mu.RUnlock()

	if !ok {
		return FailureResult("Tool '" + invocation.Name + "' not found")
	}

	return handler(ctx, invocation)
}

// Unregister removes a tool
func (r *ToolRegistry) Unregister(name string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.tools, name)
	delete(r.handlers, name)
}

// Has checks if a tool is registered
func (r *ToolRegistry) Has(name string) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	_, ok := r.tools[name]
	return ok
}

// Names returns all tool names
func (r *ToolRegistry) Names() []string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	
	result := make([]string, 0, len(r.tools))
	for name := range r.tools {
		result = append(result, name)
	}
	return result
}

// ToolParametersBuilder helps construct tool parameters
type ToolParametersBuilder struct {
	properties map[string]map[string]interface{}
	required   []string
}

// NewToolParameters creates a new parameters builder
func NewToolParameters() *ToolParametersBuilder {
	return &ToolParametersBuilder{
		properties: make(map[string]map[string]interface{}),
		required:   make([]string, 0),
	}
}

// AddString adds a string parameter
func (b *ToolParametersBuilder) AddString(name, description string) *ToolParametersBuilder {
	b.properties[name] = map[string]interface{}{
		"type":        "string",
		"description": description,
	}
	return b
}

// AddNumber adds a number parameter
func (b *ToolParametersBuilder) AddNumber(name, description string) *ToolParametersBuilder {
	b.properties[name] = map[string]interface{}{
		"type":        "number",
		"description": description,
	}
	return b
}

// AddInteger adds an integer parameter
func (b *ToolParametersBuilder) AddInteger(name, description string) *ToolParametersBuilder {
	b.properties[name] = map[string]interface{}{
		"type":        "integer",
		"description": description,
	}
	return b
}

// AddBoolean adds a boolean parameter
func (b *ToolParametersBuilder) AddBoolean(name, description string) *ToolParametersBuilder {
	b.properties[name] = map[string]interface{}{
		"type":        "boolean",
		"description": description,
	}
	return b
}

// AddEnum adds an enum parameter
func (b *ToolParametersBuilder) AddEnum(name, description string, values []string) *ToolParametersBuilder {
	b.properties[name] = map[string]interface{}{
		"type":        "string",
		"description": description,
		"enum":        values,
	}
	return b
}

// Required marks parameters as required
func (b *ToolParametersBuilder) Required(fields ...string) *ToolParametersBuilder {
	b.required = append(b.required, fields...)
	return b
}

// Build creates the parameters JSON
func (b *ToolParametersBuilder) Build() json.RawMessage {
	result := map[string]interface{}{
		"properties": b.properties,
		"required":   b.required,
	}
	data, _ := json.Marshal(result)
	return data
}

// BuildMap returns parameters as a map
func (b *ToolParametersBuilder) BuildMap() map[string]interface{} {
	return map[string]interface{}{
		"properties": b.properties,
		"required":   b.required,
	}
}

// DefineTool is a helper to create a tool with parameters builder
func DefineTool(name, description string, params *ToolParametersBuilder) Tool {
	var paramsJSON json.RawMessage
	if params != nil {
		paramsJSON = params.Build()
	}
	return Tool{
		Name:        name,
		Description: description,
		Parameters:  paramsJSON,
	}
}

File: src/typescript/src/tools.ts
/**
 * GeminiSDK Tools - Tool definition utilities.
 */

import {
  Tool,
  ToolHandler,
  ToolInvocation,
  ToolResult,
} from './types.js';

const TYPE_MAPPING: Record<string, string> = {
  string: 'string',
  number: 'number',
  boolean: 'boolean',
  object: 'object',
  array: 'array',
};

export interface ToolDefinitionOptions {
  name?: string;
  description?: string;
  parameters?: Record<string, unknown>;
}

/**
 * Define a tool for use with Gemini models.
 */
export function defineTool<T extends Record<string, unknown>>(
  options: ToolDefinitionOptions,
  handler: (args: T) => unknown | Promise<unknown>
): Tool {
  const toolName = options.name ?? handler.name ?? 'unnamed_tool';
  const description = options.description ?? `Tool: ${toolName}`;
  const parameters = options.parameters ?? { type: 'object', properties: {} };

  const wrappedHandler: ToolHandler = async (invocation: ToolInvocation): Promise<ToolResult> => {
    const result = await Promise.resolve(handler(invocation.arguments as T));

    if (typeof result === 'object' && result !== null && 'textResultForLlm' in result) {
      return result as ToolResult;
    }

    return { textResultForLlm: String(result) };
  };

  return {
    name: toolName,
    description,
    parameters,
    handler: wrappedHandler,
  };
}

/**
 * Create a tool programmatically.
 */
export function createTool(
  name: string,
  description: string,
  parameters?: Record<string, unknown>,
  handler?: ToolHandler
): Tool {
  return {
    name,
    description,
    parameters: parameters ?? { type: 'object', properties: {} },
    handler,
  };
}

/**
 * Normalize tool specifications into Tool objects.
 *
 * Accepts fully-formed Tool objects and declarative dict-style specs used by
 * gemini-cli built-in tools (e.g. `{ googleSearch: {} }`). Dict entries are
 * converted into Tool objects with no handler; the tool name is used as the
 * key and the value (if a non-empty object) is stored as the parameters schema.
 *
 * @param specs Array of Tool objects or declarative dict specs.
 * @returns Array of normalized Tool objects.
 */
export function normalizeTools(specs: Array<Tool | Record<string, unknown>>): Tool[] {
  return specs.flatMap((spec) => {
    if (typeof (spec as Tool).name === 'string' && typeof (spec as Tool).description === 'string') {
      return [spec as Tool];
    }
    return Object.entries(spec).map(([name, params]) =>
      createTool(
        name,
        `Tool: ${name}`,
        typeof params === 'object' && params !== null
          ? (params as Record<string, unknown>)
          : {},
      )
    );
  });
}

/**
 * Registry for managing tools.
 */
export class ToolRegistry {
  private tools: Map<string, Tool> = new Map();
  private categories: Map<string, Set<string>> = new Map();

  public register(tool: Tool, category?: string): void {
    this.tools.set(tool.name, tool);

    if (category) {
      if (!this.categories.has(category)) {
        this.categories.set(category, new Set());
      }
      this.categories.get(category)!.add(tool.name);
    }
  }

  public unregister(name: string): void {
    this.tools.delete(name);
    for (const categoryTools of this.categories.values()) {
      categoryTools.delete(name);
    }
  }

  public get(name: string): Tool | undefined {
    return this.tools.get(name);
  }

  public getAll(): Tool[] {
    return Array.from(this.tools.values());
  }

  public getByCategory(category: string): Tool[] {
    const toolNames = this.categories.get(category);
    if (!toolNames) return [];
    return Array.from(toolNames)
      .map((name) => this.tools.get(name))
      .filter((tool): tool is Tool => tool !== undefined);
  }

  public listCategories(): string[] {
    return Array.from(this.categories.keys());
  }
}

// Global default registry
const defaultRegistry = new ToolRegistry();

export function getDefaultRegistry(): ToolRegistry {
  return defaultRegistry;
}

export function registerTool(tool: Tool, category?: string): void {
  defaultRegistry.register(tool, category);
}
